Boss room shuffle

dewiniaid opened the pull request:
This is an unfinished work-in-progress boss shuffle feature.  Currently working:

- Dungeon boss rooms can be swapped between dungeons _(e.g. walking through the Fire Temple boss door can put you in King Dodongo's room)_
- Glitchless normal logic is adjusted to account for separation of reaching a boss room vs. defeating the boss.
- Save/quit and dying from a shuffled boss room will correctly put you at the correct dungeon (If Barinade is in Deku Tree, save/quit from Barinade will put you in Deku Tree)
- Medallion rewards are correctly distributed.  (If Fire Temple is Water Medallion, this is true regardless of which boss is at Fire Temple)
- Relocated boss rooms should be counted as part of their new dungeon for all relevant purposes (such as hints)
- Exiting a boss room (possible with Gohma and King Dodongo) _usually_ takes you to the correct location, see Known Issues below.

**Known issues:**

- MQ logic is only lightly tested.
- Glitch logic doesn't exist yet, so this feature is simply unavailable in glitch logic seeds.
- Exiting the Water Temple boss room puts you at the beginning of Water Temple rather than outside the boss room.  Unlike all other dungeons, Water Temple lacks an appropriate second entrance point at the boss door; we'd need a way to patch in another entrance to Water Temple to fix this correctly; this is outside my skillset.
- Beating Morpha puts you at the Lake Hylia water temple entrance when dungeons aren't shuffled, and presumably whereever water temple is if they are.
- Boss entrances show decoupled in the spoiler log if decoupled entrances are on.

dewiniaid committed:
Merge tag '6.0.51' into Dev-R

# Conflicts:
#	version.py

dewiniaid committed:
Merge branch 'Dev-R' of https://github.com/Roman971/OoT-Randomizer into Dev-R

# Conflicts:
#	version.py

dewiniaid committed:
Add settings for boss shuffle.

dewiniaid committed:
Add first phase of logic updates for boss shuffle.  MQ adjustments to follow.

dewiniaid committed:
Add boss shuffle entrance shuffling.

dewiniaid committed:
Finish first pass of Boss Shuffle.

dewiniaid committed:
Fix up boss shuffle addresses

This fixes errors in Deku Tree Boss Exit, Forest Temple Boss Entrance, and Shadow Temple.

It also removes addresses related to save/quit from the dungeon itself.

dewiniaid committed:
Add unique scene names to all boss rooms

Without this, Entrance Shuffle considers them part of the same scene
as the rest of the dungeon which prohibits them from rolling vanilla
(or being plando'd vanilla)

dewiniaid committed:
Fix Jabu Jabu elevator logic.

dewiniaid committed:
Boss shuffle fixes.

dewiniaid committed:
Fix water temple commit logic.

dewiniaid committed:
Merge remote-tracking branch 'roman/Dev-R' into feature/boss-shuffle

Also updates references from world.boss_shuffle to world.settings.boss_shuffle

# Conflicts:
#	EntranceShuffle.py
#	Patches.py
#	World.py

dewiniaid committed:
Add Jabu Jabu Boss Door tricks

Only relevant in boss shuffle, but the door to Jabu's boss room can be opened with the slingshot or bow from a specific location.  It can also be opened with a bombchu with the correct setup.  This adds tricks for both of those options and adjusts logic accordingly.

dewiniaid committed:
Fix decoupled entrances failing to generate when combined with boss shuffle.

Note that boss entrances themselves are not decoupled nor mixed regardless of the settings chosen.

dewiniaid committed:
Fix phantom ganon as child crash due to deku sprout CS

dewiniaid committed:
Fix(?) mixed entrances in conjunction with boss shuffle.

Also fix an issue with mixed entrance pools set to 'indoor' that was
referencing attributes on world rather than world.settings

dewiniaid committed:
Hopefully this fixes altar hints for boss shuffle.

dewiniaid committed:
Logic updates for boss shuffle, child dungeons.  Glitch logic is still unchanged.

General:
- Rename regions for clearer spoiler logs:
  - "Deku Tree Boss Room" -> "Gohma Boss Room"
  - "Jabu Jabus Belly Boss Area" -> "Jabu Jabus Belly Before Boss"
  - "Jabu Jabus Belly Boss Arena" -> "Barinade Boss Room"
  - "Dodongos Cavern Boss Area" -> "Dodongos Cavern Before Boss"
  - "Dododong Cavern Boss Arena" -> "King Dodongo Boss Room"


Deku Tree:

- Split Deku Tree Basement Ledge out from Deku Tree Lobby to mirror MQ logic
- Fix shield being required for Gohma rather than being required to reach the boss room
- MQ: Add boss shuffle logic

Dodongos Cavern:
- MQ: Add boss shuffle logic.

Jabu Jabu's Belly:

- MQ: Add boss shuffle logic
- MQ: Move the "Before Boss" area to be accessed from Jabu Jabus Belly Main,
  but only after lowered floor event is triggered (mirroring normal logic)

dewiniaid committed:
- Patch Jabu-Jabu scene to put the boss exit at a sane location.

dewiniaid committed:
Remove accidental second Deku Tree Clear event from MQ logic

dewiniaid committed:
Convert Closed Forest to Closed Deku when boss shuffle is on

In reality, we COULD probably fix this to check for the correct boss clear, but this works for now and has precedent.

dewiniaid committed:
Child logic fixes:

- Gohma properly requires a means to stun.
- Using bow or hookshot to stun Gohma is now a trick
- Actually add King Dodongo exit from MQ DC

dewiniaid committed:
Adult MQ Logic and other adjustments.

- Rename all remaining boss rooms to <boss name> Boss Room rather than <temple name> Boss Arena
- Properly require FTR or Goron Tunic to reach Fire Temple boss room
- Add Fire Temple MQ boss door logic
- Add Forest Temple MQ boss door logic
- Add Shadow Temple MQ boss door logic
- Add Spirit Temple MQ boss door logic
- Add Water Temple MQ boss door logic
- Normal Water: Remove "Water Temple Highest Water Level" region and fold
  its locations and exits into Water Temple Lobby.

dewiniaid committed:
Fix missing boss key logic for water temple.  Not certain how I accidentally removed it.

dewiniaid committed:
Logic fixes.

dewiniaid committed:
Additional Deku logic fixes.

dewiniaid committed:
Move boss rooms to Overworld.json

Boss logic is identical between normal and MQ dungeons, as boss rooms are
not affected by MQ.  Rather than duplicate the logic between both sets of
files, move the boss logic into Overworld.json per @r0bd0g's suggestion.

dewiniaid committed:
Update and rename Gohma trick to include boomerang as a possibility.

dewiniaid committed:
Fix text on boss shuffle; Deku Tree has been included for quite some time.

I could have sworn I already did this.

dewiniaid committed:
Shuffle bosses before dungeon entrances rather than after.

This drastically improves the success rate (and thus time required) for seed generation, particularly in multiworlds.

SlyryD commented on data/World/Jabu Jabus Belly MQ.json:
```diff
@@ -44,23 +45,24 @@
     },
     {
-        "region_name": "Jabu Jabus Belly Boss Area",
+        "region_name": "Jabu Jabus Belly Before Boss",
```
Just FYI, this region is referring to anything after big octo. The cow is in the room after fighting big octo.

SlyryD replied:
Probably doesn't change the naming at all, but I thought I'd let you know.

dewiniaid committed:
Fix fire temple boss door logic.

This was previously fixed in shortcuts branches, but we missed it in non-shortcuts somehow.

dewiniaid committed:
Change most boss exit conditions from False to True.

This works around a convoluted series of issues that ultimately causes
boss shuffle to be incompatible with Python 3.6 and Python 3.7, which we
claim to support.

While the changed logic is technically wrong, it has no effect as the
mentioned locations are always accessible if the boss room itself is.  This
will probably need to be fixed in the future if decoupled bosses are ever
added, but that has so many other logic implications I think we can ignore
it for now.

dewiniaid committed:
Speculative fix for Morpha blue warp in boss shuffle.

dewiniaid commented:
I forgot I had this PR over here.

Closing this, as the new plan is to have it merged into Dev and osmosis its way over here.  I am still internally maintaining a Boss Shuffle + Dev-R fork to account for the differences in ER in this branch (namely decoupled/mixed, which should exclude shuffled bosses).
